### PR TITLE
Wait for completion of the upload queue in flush_frozen_layer

### LINKED ISSUE #7317: pageserver: backpressure on layer freeze/flush
Currently the flush loop can flush any number of layers: they get enqueued for upload but we don't wait for the upload to complete.

If S3 uploads are slower than we are ingesting data, then we can build up rather large buffers of flushed-but-not-uploaded layers: more data than we can upload within our clean shutdown time budget, and so much data that the next startup is quite stressful as we try to re-ingest it all.

We may apply some simple backpressure by adding a wait in the flush loop: immediately after we schedule the index upload, we may wait for the remote timeline client to catch up.

The resulting buffer size for each tenant would be two layers: one in-memory layer they're writing to, and one frozen/flushed layer that is waiting to upload to S3.

### PR DESCRIPTION
**Pull request by arpad-m**
Makes `flush_frozen_layer` add a barrier to the upload queue and makes it wait for that barrier to be reached until it lets the flushing be completed.

This gives us backpressure and ensures that writes can't build up in an unbounded fashion.

Fixes #7317

### CODE REVIEW

**Review by koivunej — APPROVED** [review]
This is a valid implementation for the issue.

I am wondering a bit about the benchmarks. Could you push an empty commit so we see if anything breaks so we don't need to fix it after merging?

I am so wondering if this should have a test.. In theory it largely exercises existing wal backpressure tests, however, there we never pause the upload queue (intentionally at least). Did you try to delay those, did anything break?

**Comment by arpad-m** [comment]
Given the breakages of the tests, I suppose I'll need to modify them already. Maybe this will be sufficient already and no further testing will be needed?

File: pageserver/src/tenant/timeline.rs [hunk 1]
```diff
@@ -4077,7 +4080,18 @@ impl Timeline {
...
+            .await
+            .map_err(|e| match e {
+                WaitCompletionError::UploadQueueShutDownOrStopped
+                | WaitCompletionError::NotInitialized(
+                    NotInitialized::ShuttingDown | NotInitialized::Stopped,
+                ) => FlushLayerError::Cancelled,
+                WaitCompletionError::NotInitialized(NotInitialized::Uninitialized) => {
+                    FlushLayerError::Other(anyhow!(e).into())
+                }
+            })?;
```
Comment (koivunej) [inline pageserver/src/tenant/timeline.rs:4094]: I would had thought there was a conversion `From<WaitCompletionError> for FlushLayerError`? maybe it was another map_err?

**Comment by koivunej** [comment]
> Maybe this will be sufficient already and no further testing will be needed?

Might be for example `test_compaction_waits_for_upload` becomes one. It certainly cannot work as it did...

Pretty much all tests using `"before-upload-layer-pausable"` should now be broken.

File: pageserver/src/tenant/timeline.rs (hunk 1 above)
Comment (arpad-m) [reply pageserver/src/tenant/timeline.rs:4094]: There is no conversion, if I remove the map_err it gives an error.

**Comment by arpad-m** [comment]
List of failing tests:

  *  test_cannot_branch_from_non_uploaded_branch: [release](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-8550/10163048773/index.html#suites/ffbb7f9930a77115316b58ff32b7c719/a441ea1fae6d6e42), [debug](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-8550/10163048773/index.html#suites/ffbb7f9930a77115316b58ff32b7c719/b2181d581e9203c)
  *  test_cannot_create_endpoint_on_non_uploaded_timeline: [release](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-8550/10163048773/index.html#suites/ffbb7f9930a77115316b58ff32b7c719/b793a51cb596f345), [debug](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-8550/10163048773/index.html#suites/ffbb7f9930a77115316b58ff32b7c719/41105dd9cf89edfd)
  *  test_remote_storage_upload_queue_retries: [release](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-8550/10163048773/index.html#suites/837740b64a53e769572c4ed7b7a7eeeb/bb026a9c133eade0), [debug](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-8550/10163048773/index.html#suites/837740b64a53e769572c4ed7b7a7eeeb/da6f1f462cd2e3fc)
  *  test_compaction_waits_for_upload: [release](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-8550/10163048773/index.html#suites/837740b64a53e769572c4ed7b7a7eeeb/6fbab31e801e4baf), [debug](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-8550/10163048773/index.html#suites/837740b64a53e769572c4ed7b7a7eeeb/b5a06fd7dde077f)

**Review by tristan957 — APPROVED** [review]
Looks good on compute side, except maybe s/NEON_COMPUTE_TESTING_BASEBACKUP_TRIES/NEON_COMPUTE_TESTING_BASEBACKUP_RETRIES. Up to you.

**Comment by arpad-m** [comment]
Hmmm benchmarks fail in CI ([link a](https://github.com/neondatabase/neon/actions/runs/10200589667/job/28222942955?pr=8550), [link b](https://github.com/neondatabase/neon/actions/runs/10200589667/job/28222944067?pr=8550)).

```
 FAILED test_runner/performance/test_gc_feedback.py::test_gc_feedback_with_snapshots[release-pg14-github-actions-selfhosted] - Failed: Timeout >10000.0s
 
 
FAILED test_runner/performance/test_gc_feedback.py::test_gc_feedback[release-pg14-github-actions-selfhosted] - Failed: Timeout >10000.0s
```

**Comment by koivunej** [comment]
> Hmmm benchmarks fail in CI ([link a](https://github.com/neondatabase/neon/actions/runs/10200589667/job/28222942955?pr=8550), [link b](https://github.com/neondatabase/neon/actions/runs/10200589667/job/28222944067?pr=8550)).
> 
> ```
>  FAILED test_runner/performance/test_gc_feedback.py::test_gc_feedback_with_snapshots[release-pg14-github-actions-selfhosted] - Failed: Timeout >10000.0s
>  
>  
> FAILED test_runner/performance/test_gc_feedback.py::test_gc_feedback[release-pg14-github-actions-selfhosted] - Failed: Timeout >10000.0s
> ```

These failed on main.

**Pull request merged**